ledger/catchpointtracker.go

```diff
@@ -0,0 +1,908 @@
+
+const (
+	// trieRebuildAccountChunkSize defines the number of accounts that would get read at a single chunk
+	// before added to the trie during trie construction
+	trieRebuildAccountChunkSize = 16384
+	// trieRebuildCommitFrequency defines the number of accounts that would get added before we call evict to commit the changes and adjust the memory cache.
+	trieRebuildCommitFrequency = 65536
+	// trieAccumulatedChangesFlush defines the number of pending changes that would be applied to the merkle trie before
+	// we attempt to commit them to disk while writing a batch of rounds balances to disk.
+	trieAccumulatedChangesFlush = 256
+)
+
+// TrieMemoryConfig is the memory configuration setup used for the merkle trie.
+var TrieMemoryConfig = merkletrie.MemoryConfig{
+	NodesCountPerPage:         merkleCommitterNodesPerPage,
+	CachedNodesCount:          trieCachedNodesCount,
+	PageFillFactor:            0.95,
+	MaxChildrenPagesThreshold: 64,
+}
+
+type catchpointTracker struct {
+	// dbDirectory is the directory where the ledger and block sql file resides as well as the parent directory for the catchup files to be generated
+	dbDirectory string
+
+	// catchpointInterval is the configured interval at which the accountUpdates would generate catchpoint labels and catchpoint files.
+	catchpointInterval uint64
+
+	// catchpointFileHistoryLength defines how many catchpoint files we want to store back.
+	// 0 means don't store any, -1 mean unlimited and positive number suggest the number of most recent catchpoint files.
+	catchpointFileHistoryLength int
+
+	// archivalLedger determines whether the associated ledger was configured as archival ledger or not.
+	archivalLedger bool
+
+	// Prepared SQL statements for fast accounts DB lookups.
+	accountsq *accountsDbQueries
+
+	// log copied from ledger
+	log logging.Logger
+
+	// Connection to the database.
+	dbs db.Pair
+
+	// The last catchpoint label that was written to the database. Should always align with what's in the database.
+	// note that this is the last catchpoint *label* and not the catchpoint file.
+	lastCatchpointLabel string
+
+	// catchpointSlowWriting suggest to the accounts writer that it should finish writing up the catchpoint file ASAP.
+	// when this channel is closed, the accounts writer would try and complete the writing as soon as possible.
+	// otherwise, it would take it's time and perform periodic sleeps between chunks processing.
+	catchpointSlowWriting chan struct{}
+
+	// catchpointWriting help to synchronize the catchpoint file writing. When this atomic variable is 0, no writing is going on.
+	// Any non-zero value indicates a catchpoint being written.
+	catchpointWriting int32
+
+	// The Trie tracking the current account balances. Always matches the balances that were
+	// written to the database.
+	balancesTrie *merkletrie.Trie
+
+	// accountsMu is the synchronization mutex for accessing the various non-static variables.
+	accountsMu deadlock.RWMutex
```

Comment:
Maybe give this a different name like `catchpointsMu` or something so it doesn't get confused with `accountUpdates.accountsMu`?